In [ ]:
# Sentiment Analysis with Naive Bayes (Optimized for Efficiency)
# Name: MUHAMMAD IMAN ARIF BIN MAUZI
# Student ID: SW01083215

# Name: MUHAMMAD 'UMAR BIN ZOLKIFLE
# Student ID: SW01082397

import pandas as pd
import re
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score
from sklearn.pipeline import make_pipeline
import nltk
from nltk.corpus import stopwords

# Download only once
nltk.download('stopwords', quiet=True)

# ----------------------
# 1. Efficient Data Loading
# ----------------------
# Load only necessary columns
cols = ['Text', 'Score']
dtypes = {'Text': 'string', 'Score': 'int8'}
df = pd.read_csv('Reviews_Sample.csv', usecols=cols, dtype=dtypes)

# Filter and binarize sentiment
df = df[df['Score'] != 3].copy()
df['Sentiment'] = (df['Score'] > 3).astype('int8')

# ----------------------
# 2. Optimized Preprocessing
# ----------------------
# Precompile regex and stopwords
stop_words = set(stopwords.words('english'))
pattern = re.compile(r'[^a-zA-Z\s]')

def fast_clean(text):
    # Combined operations for efficiency
    return ' '.join([
        word for word in 
        pattern.sub('', text.lower()).split() 
        if word not in stop_words and len(word) > 2
    ])

# Vectorized cleaning
df['Clean_Text'] = df['Text'].apply(fast_clean)

# ----------------------
# 3. Efficient Feature Extraction & Model
# ----------------------
# Pipeline with TF-IDF and NB
model = make_pipeline(
    TfidfVectorizer(max_features=2500, sublinear_tf=True),
    MultinomialNB(alpha=0.1)
)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    df['Clean_Text'], df['Sentiment'], 
    test_size=0.2, random_state=42, stratify=df['Sentiment']
)

# Train and predict
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

# ----------------------
# 4. Evaluation
# ----------------------
print(f"Accuracy: {accuracy_score(y_test, y_pred):.2f}")
print(f"F1-Score: {f1_score(y_test, y_pred):.2f}")

# ----------------------
# Discussion (Add to script)
# ----------------------
# Strengths:
# - 10x faster training than SVM
# - Handles high dimensionality efficiently
# - Minimal memory requirements
# - Naturally handles small datasets

# Weaknesses:
# - Assumes feature independence
# - Struggles with phrase-based sentiment
# - Cannot handle unknown words gracefully

Accuracy: 0.88
F1-Score: 0.93


In [ ]:
# Sentiment Analysis with Logistic Regression
# Name: MUHAMMAD IMAN ARIF BIN MAUZI
# Student ID: SW01083215

# Name: MUHAMMAD 'UMAR BIN ZOLKIFLE
# Student ID: SW01082397

import pandas as pd
import re
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score
from sklearn.pipeline import make_pipeline

# ----------------------
# 1. Efficient Data Loading
# ----------------------
df = pd.read_csv('Reviews_Sample.csv', usecols=['Text', 'Score'], dtype={'Text': 'string', 'Score': 'int8'})
df = df[df['Score'] != 3].copy()
df['Sentiment'] = (df['Score'] > 3).astype('int8')

# ----------------------
# 2. Optimized Preprocessing
# ----------------------
stop_words = set(pd.read_csv('https://raw.githubusercontent.com/stopwords-iso/stopwords-en/master/stopwords-en.txt', 
                           header=None, names=['words']).words.values)  # External stopwords

df['Clean_Text'] = df['Text'].str.lower() \
                             .str.replace(r'[^a-z\s]', '', regex=True) \
                             .apply(lambda x: ' '.join([w for w in x.split() if w not in stop_words and len(w) > 2]))

# ----------------------
# 3. Model Pipeline
# ----------------------
model = make_pipeline(
    TfidfVectorizer(max_features=3000, sublinear_tf=True),
    LogisticRegression(class_weight='balanced', max_iter=1000, solver='saga')
)

X_train, X_test, y_train, y_test = train_test_split(
    df['Clean_Text'], df['Sentiment'], 
    test_size=0.2, random_state=42
)

model.fit(X_train, y_train)
y_pred = model.predict(X_test)

# ----------------------
# 4. Evaluation & Discussion
# ----------------------
print(f"Accuracy: {accuracy_score(y_test, y_pred):.2f}")
print(f"F1-Score: {f1_score(y_test, y_pred):.2f}")

# Strengths:
# - Better handles class imbalance than NB
# - Provides feature importance
# - Faster convergence than SVM

# Weaknesses:
# - Requires regularization tuning
# - Linear decision boundaries

Accuracy: 0.87
F1-Score: 0.92


In [ ]:
# Sentiment Analysis with Random Forest
# Name: MUHAMMAD IMAN ARIF BIN MAUZI
# Student ID: SW01083215

# Name: MUHAMMAD 'UMAR BIN ZOLKIFLE
# Student ID: SW01082397

import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_extraction.text import HashingVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score

# ----------------------
# 1. Memory-Optimized Loading
# ----------------------
df = pd.read_csv('Reviews_Sample.csv', usecols=['Text', 'Score'])
df = df[df['Score'] != 3]
df['Sentiment'] = (df['Score'] > 3).astype(int)
df['Text'] = df['Text'].str.lower().str.replace(r'[^a-z\s]', '', regex=True)

# ----------------------
# 2. Streamlined Processing
# ----------------------
# HashingVectorizer for memory efficiency
vectorizer = HashingVectorizer(n_features=2**16, alternate_sign=False)

X = vectorizer.fit_transform(df['Text'])
y = df['Sentiment']

# ----------------------
# 3. Optimized Random Forest
# ----------------------
model = RandomForestClassifier(n_estimators=50, 
                              max_depth=15, 
                              class_weight='balanced',
                              n_jobs=-1)  # Parallel processing

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

model.fit(X_train, y_train)
y_pred = model.predict(X_test)

# ----------------------
# 4. Evaluation & Discussion
# ----------------------
print(f"Accuracy: {accuracy_score(y_test, y_pred):.2f}")
print(f"F1-Score: {f1_score(y_test, y_pred):.2f}")

# Strengths:
# - Handles non-linear relationships
# - Robust to outliers
# - No need for TF-IDF preprocessing

# Weaknesses:
# - Slower prediction time
# - Higher memory usage
# - Less interpretable

Accuracy: 0.85
F1-Score: 0.91


<h1>Model Comparison and Analysis</h1>

<h2>Based on the results:</h2>
<table>
    <caption>Table 1</caption>
    <thead>
        <tr>
            <th>Model</th>
            <th>Accuracy</th>
            <th>F1 Score</th>
            <th>Training Speed </th>
            <th>Interpretability</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                <p>Naive Bayes</p>
            </td>
            <td>0.88</td>
            <td>0.93</td>
            <td>Fastest</td>
            <td>Moderate</td>
        </tr>
        <tr>
            <td>
                <p>Logistic Regression</p>
            </td>
            <td>0.87</td>
            <td>0.92</td>
            <td>Fast</td>
            <td>High</td>
        </tr>
        <tr>
            <td>
                <p>Random Forest</p>
            </td>
            <td>0.84</td>
            <td>0.90</td>
            <td>Slowest</td>
            <td>Low</td>
        </tr>
    </tbody>
</table>

<h2>Key Observations</h2>

<h3>Performance:</h3>

Naive Bayes outperforms others in both metrics, likely due to:
- Efficient handling of high-dimensional TF-IDF features
- Suitability for text classification tasks

Logistic Regression follows closely, showing linear models work well for sentiment analysis

Random Forest underperforms due to:
- Difficulty with sparse text data
- Overfitting on noisy features